README.md
# About MULTIPASS

MULtiple TIme constant Psd scintillATion emiSSion example - MULTIPASS, 
Is an example code demonstrating scintillation in Geant4 with multiple time constants for use with scintillators which exhibit pulse shape discrimination (PSD). It is based off of the code for simulating the EJ-276 detector system in Michael Hubbard’s PhD thesis.

# Building
The MULTIPASS source code has been built and tested against Gean4.10.04.p01 other versions may work.
It is suggested you make a build directory for example ```build_MULTIPASS.```
Inside the build directory run:
```bash
cmake -DGeant4_DIR=path/to/geant4/install/directory path/to/MULTIPASS
```
then
```bash
make
```
this will produce an executable called ```MULTIPASS```

# Running
```
./MULTIPASS --help
  - a simulation software for MULTIPASS detector
  Usage:
  ./MULTIPASS [OPTION...]
  -m, --macro arg run macro
  -r, --runmode arg Run mode [Full, Gamma, Optics] (default: Optics)
  -h, --help Print help
```

# Data output


ChangeLog.md
# Change Log
All notable changes to this project will be documented in this file.

## [Unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security

Added for new features.
Changed for changes in existing functionality.
Deprecated for once-stable features removed in upcoming releases.
Removed for deprecated features removed in this release.
Fixed for any bug fixes.
Security to invite users to upgrade in case of vulnerabilities.

Dates: YYYY-MM-DD https://xkcd.com/1179/

=====================================================================
## Known Bugs

### Double counting in MT mode:
 - Reflections are double, triple etc counted unknown cause.

### Vis issue in batch mode
 - WARNING: G4VisManager::IsValidView(): Attempt to draw when no graphics system
  has been instantiated.  Use "/vis/open" or "/vis/sceneHandler/create".
  Alternatively, to avoid this message, suppress instantiation of vis
  manager (G4VisExecutive) and ensure drawing code is executed only if
  G4VVisManager::GetConcreteInstance() is non-zero.

### No onscreen visualisation in MT mode
 - Particle tracks etc. do not appear in MT mode, they do in
   sequential mode. Observed in 10.4.1 and 10.3.1 etc.

### EventAction reconstruction:
 - Currently list position of interaction at (0,0,0) this is incorrect due to 
   incorrect passing of PMT position in PMTSD.

### LXE Bugs
 - Incorrect #include gaurd in LXeWLSFiber.hh (Fixed on bugzilla)
 - Incorrect naming of G4LogicalSkinSurface("housing_surf") in MainVolume 
   it is set to "photocath_surf" which is incorrect
 - Position reconstruction in eventAction also does not work.

### Seg fault
 - boundaryStatus=boundary->GetStatus();
 - when run in Gamma runmode. Might be caused by optical photon generated in PGA

=====================================================================
## ToDos
 - 

=====================================================================
### Notes:
 - Gammas event information increase escape energy may now be wrong as 
   gammas are no longer just killed

  - MT SEGFAULTS trying to use get functions for primaries no idea why

=====================================================================
## 2020-04-02 
### Added 
### Changed
### Deprecated
### Removed
 - Code clean up completed
### Fixed
### Security

## 2020-03-18 
### Added 
### Changed
 - EJ276 code is now MULTIPASS example code

### Deprecated
### Removed
### Fixed
### Security




MULTIPASS.cc
//
// File:         MULTIPASS/MULTIPASS.cc
// Description:  Main program for MULTIPASS simulation. Based on original code 
//               from Geant4 optical/LXe from 10.4.p01
// Version:      1.0.0
// Created:      2020-03-18
// Author:       Michael Hubbard
// Modified:     
//
//
// Tested with:
//              Geant4 v 10.4.p01
//              ROOT v 6.12.06
//
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//

// MT doesn't appear to work with this example
// Therefore uses regular sequential mode
#include "G4RunManager.hh"


#include "G4UImanager.hh"
#include "G4String.hh"

#include "MULTIPASSPhysicsList.hh"
#include "MULTIPASSDetectorConstruction.hh"

#include "MULTIPASSActionInitialization.hh"

#include "Randomize.hh"
#include "time.h"

#ifdef G4VIS_USE
  #include "G4VisExecutive.hh"
#endif

#ifdef G4UI_USE
  #include "G4UIExecutive.hh"
#endif

#include "cxxopts.hpp"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

int main(int argc, char** argv)
{

  // Note on use of cxx:
  // Options are declared with a long and an optional short option. 
  // A description must be provided. 
  // The third argument is the value, if omitted it is boolean. 
  // Any type can be given as long as it can be parsed, with operator>>.
  cxxopts::Options options(argv[0],
                        " - a simulation software for MULTIPASS detector");
  options.positional_help("[optional args]")
      .show_positional_help();

  options.add_options()
    ("m,macro", "run macro", cxxopts::value<std::string>())
    ("r,runmode", "Run mode [Full, Gamma, Optics]", 
                         cxxopts::value<std::string>()->default_value("Optics"))
    ("h,help", "Print help")
    ;

  auto result = options.parse(argc, argv);
  if (result.count("help"))
  {
    G4cout << options.help({""}) << G4endl;
    exit(0);
  }
  // If macro detected run in batch mode
  G4bool interactiveMode = not(result.count("macro") > 0);


  // Run mode options

  G4String runMode = result["runmode"].as<std::string>();

  std::vector<G4String> runOpt = {"Full", "Gamma", "Optics"};

  G4bool runOptOK = false;

  for (const G4String& i : runOpt) { // access by const reference
    if (runMode==i ) runOptOK = true;
  }
  if (not runOptOK) {
    G4cout << runMode << " is not a valid runmode" << G4endl;
    G4cout << "Valid options: ";
    G4String delim = "";
    for (const G4String& i : runOpt) { // access by const reference
      G4cout << delim << i;
      delim = ", ";
    }
    G4cout << G4endl; 
    exit(1);
  }

  // If mode is Gamma optical physics is disabled
  G4bool enableOptics = (runMode!="Gamma") ; 


  //set random seed with system time
  G4long myseed = time(NULL);

  // Choose the Random engine
  G4Random::setTheEngine(new CLHEP::RanecuEngine);
  // Seed the random number generator manually
  G4Random::setTheSeed(myseed);

  G4RunManager * runManager = new G4RunManager;

  runManager->SetUserInitialization(
                              new MULTIPASSDetectorConstruction());
  runManager->SetUserInitialization(
                              new MULTIPASSPhysicsList(enableOptics));

  runManager->SetUserInitialization(
                              new MULTIPASSActionInitialization(runMode));


  runManager->Initialize();
  
#ifdef G4VIS_USE
  G4VisManager* visManager = new G4VisExecutive;
  // G4VisExecutive can take a verbosity argument - see /vis/verbose guidance.
  // G4VisManager* visManager = new G4VisExecutive("Quiet");
  visManager->Initialize();
#endif


 
  // get the pointer to the UI manager and set verbosities
  G4UImanager* UImanager = G4UImanager::GetUIpointer();

  if(interactiveMode){
#ifdef G4UI_USE
    G4UIExecutive* ui = new G4UIExecutive(argc, argv);
#ifdef G4VIS_USE
    UImanager->ApplyCommand("/control/execute vis.mac");
#endif
    if (ui->IsGUI())
       UImanager->ApplyCommand("/control/execute gui.mac");
    ui->SessionStart();
    delete ui;
#endif
  }
  else{
    G4String command = "/control/execute ";
    G4String filename = result["macro"].as<std::string>();
    UImanager->ApplyCommand(command+filename);
  }

#ifdef G4VIS_USE
  delete visManager;
#endif

  // job termination
  delete runManager;
  return 0;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
